### PR TITLE
Implementing musig2

### PR DESCRIPTION
**Pull request by NicolasDorier**
Typical usage:

```csharp
var msg32 = random_scalar_order().ToBytes();
var tweak = random_scalar_order().ToBytes();
var adaptor = ctx.CreateECPrivKey(random_scalar_order());
var peers = 3;
var privKeys = new ECPrivKey[peers];
var privNonces = new MusigPrivNonce[peers];
var pubNonces = new MusigPubNonce[peers];
var musig = new MusigContext[peers];
var sigs = new MusigPartialSignature[peers];
var pubKeys = new ECXOnlyPubKey[peers];

for (int i = 0; i < peers; i++)
{
	privKeys[i] = ctx.CreateECPrivKey(random_scalar_order());
	pubKeys[i] = privKeys[i].CreateXOnlyPubKey();
}

for (int i = 0; i < peers; i++)
{
	musig[i] = new MusigContext(pubKeys, msg32);
	privNonces[i] = musig[i].GenerateNonce((uint)i, privKeys[i]);
	pubNonces[i] = privNonces[i].CreatePubNonce();
}

for (int i = 0; i < peers; i++)
{
	if (useTweak)
	{
		musig[i].Tweak(tweak);
	}
	if (useAdaptor)
	{
		musig[i].UseAdaptor(adaptor.CreatePubKey());
	}

	musig[i].ProcessNonces(pubNonces);
	sigs[i] = musig[i].Sign(privKeys[i], privNonces[i]);
}


// Verify all the partial sigs
for (int i = 0; i < peers; i++)
{
	Assert.True(musig[i].Verify(pubKeys[i], pubNonces[i], sigs[i]));
}

// Combine
var schnorrSig = musig[0].Combine(sigs);

if (useAdaptor)
	schnorrSig = musig[0].Adapt(schnorrSig, adaptor);
// Verify resulting signature
// SigningPubKey is the tweaked key if tweaked, or the combined key if not
Assert.True(musig[0].SigningPubKey.SigVerifyBIP340(schnorrSig, msg32));
```

### CODE REVIEW

**Comment by dangershony** [comment]
Brilliant, the typical use example will be hard to find when this PR is closed that will be a shame.

File: NBitcoin/Secp256k1/Musig/MusigPrivNonce.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+			}
+			if (!(msg32.Length is 32) && !(msg32.Length is 0))
+				throw new ArgumentException(nameof(msg32), "msg32 must be 32 bytes or 0 bytes");
+			if (!(extraInput32.Length is 32) && !(extraInput32.Length is 0))
+				throw new ArgumentException(nameof(extraInput32), "extraInput32 must be 32 bytes or 0 bytes");
+
+			Span<byte> key32 = stackalloc byte[32];
+			if (signingKey is null)
+			{
+				key32 = key32.Slice(0, 0);
```
Comment (lontivero) [inline NBitcoin/Secp256k1/Musig/MusigPrivNonce.cs:56]: ? I don't understand what `Slice(0, 0)` does here.

Comment (NicolasDorier) [reply NBitcoin/Secp256k1/Musig/MusigPrivNonce.cs:56]: mmmh maybe I should better used nullable Span if possible.

**Comment by NicolasDorier** [comment]
@dangershony I should really work on a better doc of NBitcoin :(

File: NBitcoin/Secp256k1/Musig/MusigPrivNonce.cs (hunk 1 above)
Comment (NicolasDorier) [reply NBitcoin/Secp256k1/Musig/MusigPrivNonce.cs:56]: Can't use nullable on spans

**Review by jmayes41 — CHANGES_REQUESTED** [review]
Please review and change

**Pull request merged**